design-system/src/theme-provider.tsx

```diff
@@ -54,34 +57,50 @@ const validateTheme = (themeName?: string): ThemeName => {
   return 'default';
 };
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
-
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
-
-  useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
-    });
-  }, [theme, props.scope]);
-
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
-
-  return (
-    <ThemeContext.Provider value={value}>
-      <div ref={localScopeElement}>{props.children}</div>
-    </ThemeContext.Provider>
-  );
-};
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
+
+const ThemeProvider = forwardRef<HTMLDivElement, ThemeProviderProps>(
+  (props, ref) => {
+    const rootRef = useRef<HTMLElement>(
+      isBrowser ? document.querySelector<HTMLElement>(':root') : null
+    );
+    const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+
+    const changeTheme = useCallback((newTheme: string) => {
+      setTheme(validateTheme(newTheme));
+    }, []);
+
+    const localRef = ref as unknown as RefObject<HTMLDivElement>;
+
+    useEffect(() => {
+      const targetElement = localRef?.current
+        ? localRef.current
+        : rootRef?.current;
```

Comment:
Nit: I'm not a big fan of using refs, as it's a bit tricky to manage and type. Also here you see that we create a `rootRef` even though we might not use it.

As mentioned in our last meeting, we can also approach this like `react-modal` does: https://github.com/reactjs/react-modal/blob/403093510eae51b5ee289ee0a08e79ee1be068dd/src/components/Modal.js#L101

Overriding the target element is simply a matter of implementing the selector function.

To think about it...the goal is to make the setup simpler.

Reply:
Thanks, I got rid of refs